### PR TITLE
Update TaskGroup removal logic to work for fire-and-forget tasks

### LINKED ISSUE #3465: Error removing task group
Believe this is related to #3322: running on master since pulling in that PR, I'm suddenly seeing a lot of these errors:
```
distributed.scheduler - ERROR - Error transitioning '_plot_hourly_scatterplots-e923799215afb4e224b3d5672cdf8be3' from 'processing' to 'memory'
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/distributed/scheduler.py", line 4658, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
distributed.core - ERROR - list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/distributed/core.py", line 473, in handle_stream
    handler(**merge(extra, msg))
  File "/usr/local/lib/python3.7/site-packages/distributed/scheduler.py", line 2600, in handle_task_finished
    r = self.stimulus_task_finished(key=key, worker=worker, **msg)
  File "/usr/local/lib/python3.7/site-packages/distributed/scheduler.py", line 2015, in stimulus_task_finished
    recommendations = self.transition(key, "memory", worker=worker, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/distributed/scheduler.py", line 4658, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
```
Played around a bit with trying to simply reproduce but no luck so far; some possible culprits are:
- many of these tasks are stolen
- some workers could be starting/stopping during processing
- I believe the tasks that are failing are being `fire_and_forget`ten 

@jrbourbeau anything stand out to you...? if it's possible for `ts.group.states` and `ts.prefix.groups` to be inconsistent then maybe there should be another `if` or `try/pass` here regardless?

### LINKED ISSUE #3551: fire_and_forget scheduler error in 2.11.0
Hi,

Discovered what looks like a bug with fire_and_forget in the scheduler in 2.11.0. The error doesn't exist in 2.10.1. 

**Repro**

Running a 1-worker cluster like:
```
docker run -it --network host daskdev/dask:2.11.0 dask-scheduler
docker run -it --network host daskdev/dask:2.11.0 dask-worker localhost:8786
```

I run a python shell:
```
virtualenv .venv
source .venv/bin/activate

pip install dask distributed
python
```

Using the small dask.distributed example:
```
import time
import random
import dask.distributed
import logging

 
def inc(x):
    logging.warning('blah: x')
    time.sleep(random.random())
    return x + 1


def dec(x):
    logging.warning('blah: y')
    time.sleep(random.random())
    return x - 1


def add(x, y):
    logging.warning('blah: z')
    time.sleep(random.random())
    return x + y


def repro():
    with dask.config.set({"distributed.comm.compression": "lz4"}):
        client = dask.distributed.Client("tcp://localhost:8786")
        f = client.get({
            'x': (inc, 1),
            'y': (dec, 2),
            'z': (add, 'x', 'y')
        }, 'z', sync=False)
        dask.distributed.fire_and_forget(f)


repro()
```

I get this error in the scheduler:
```
distributed.scheduler - INFO - Plugin failed with exception
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 4645, in transition
    plugin.transition(key, start, finish2, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/distributed/diagnostics/graph_layout.py", line 93, in transition
    self.visible_updates.append((self.index[key], "False"))
KeyError: 'z'
distributed.scheduler - ERROR - Error transitioning 'z' from 'processing' to 'memory'
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 4655, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
distributed.core - ERROR - list.remove(x): x not in list
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/distributed/core.py", line 473, in handle_stream
    handler(**merge(extra, msg))
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2597, in handle_task_finished
    r = self.stimulus_task_finished(key=key, worker=worker, **msg)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2015, in stimulus_task_finished
    recommendations = self.transition(key, "memory", worker=worker, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 4655, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
distributed.scheduler - INFO - Remove worker <Worker 'tcp://127.0.0.1:36953', name: tcp://127.0.0.1:36953, memory: 0, processing: 0>
distributed.core - INFO - Removing comms to tcp://127.0.0.1:36953
distributed.scheduler - INFO - Lost all workers
distributed.utils - ERROR - list.remove(x): x not in list
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/distributed/utils.py", line 665, in log_errors
    yield
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 1739, in add_worker
    await self.handle_worker(comm=comm, worker=address)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2694, in handle_worker
    await self.handle_stream(comm=comm, extra={"worker": worker})
  File "/opt/conda/lib/python3.7/site-packages/distributed/core.py", line 473, in handle_stream
    handler(**merge(extra, msg))
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2597, in handle_task_finished
    r = self.stimulus_task_finished(key=key, worker=worker, **msg)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2015, in stimulus_task_finished
    recommendations = self.transition(key, "memory", worker=worker, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 4655, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
distributed.core - ERROR - list.remove(x): x not in list
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/distributed/core.py", line 412, in handle_comm
    result = await result
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 1739, in add_worker
    await self.handle_worker(comm=comm, worker=address)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2694, in handle_worker
    await self.handle_stream(comm=comm, extra={"worker": worker})
  File "/opt/conda/lib/python3.7/site-packages/distributed/core.py", line 473, in handle_stream
    handler(**merge(extra, msg))
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2597, in handle_task_finished
    r = self.stimulus_task_finished(key=key, worker=worker, **msg)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 2015, in stimulus_task_finished
    recommendations = self.transition(key, "memory", worker=worker, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/distributed/scheduler.py", line 4655, in transition
    ts.prefix.groups.remove(tg)
ValueError: list.remove(x): x not in list
distributed.scheduler - INFO - Register worker <Worker 'tcp://127.0.0.1:36953', name: tcp://127.0.0.1:36953, memory: 0, processing: 0>
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:36953
distributed.core - INFO - Starting established connection
```

All of the functions complete on the worker (I see the output printed). However, once the exception is thrown, the next call to `repro` does not get executed. A third call to `repro` does get executed and again throws the exception.

Running the exact same code against 2.10.1 scheduler and worker runs without the exception and behaves as expected.

Thanks,

Mark

### PR DESCRIPTION
**Pull request by jrbourbeau**
tl;dr: Currently `fire-and-forget` tasks result in nested task transitions which both end in the `forgotten` state and cause an error to be raised (xref #3465 and #3551). This PR updates the task group removal logic to work for these situations. A longer explanation follows below. 

cc @bnaul @mgh35 if you get a change to take a look

When a fired-and-forgotten task has been executed, as it's being added to the set of tasks in memory here (part of the `processing`->`memory` transition):

https://github.com/dask/distributed/blob/ae74b5ea18fac7e272c9f25b5d9f2775956aa943/distributed/scheduler.py#L3821-L3823

it is also transitioned to the `forgotten` state here

https://github.com/dask/distributed/blob/ae74b5ea18fac7e272c9f25b5d9f2775956aa943/distributed/scheduler.py#L2295

Overall this has the affect of doing a nested transition (`memory`->`forgotten` happening inside the `processing` -> `memory` transition).

Before the task completes the `processing` -> `memory` transition, it goes through the `memory`->`forgotten` transition where the task is put into the `forgotten` state and the corresponding task group is deleted. Then when the `processing` -> `memory` transition is resumed, the task has already been put into the `forgotten` state and a second attempt to delete the (already deleted) task group causes an error to be raised.

Fixes #3465, fixes #3551

### CODE REVIEW

**Comment by mgh35** [comment]
Applying this patch locally fixes the issue I was seeing. Thank you very much!

**Comment by dhirschfeld** [comment]
> *It'd be good to include this in the release today*

+:100:

**Comment by bnaul** [comment]
Looks good to me, thanks @jrbourbeau!

**Comment by jrbourbeau** [comment]
Thanks all for the feedback!
